core/src/main/scala/kafka/server/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,198 @@
+package kafka.server
+
+import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
+
+import kafka.coordinator.group.GroupCoordinator
+import kafka.metrics.KafkaMetricsGroup
+import kafka.server.QuotaFactory.QuotaManagers
+import kafka.utils.ShutdownableThread
+import org.apache.kafka.common.internals.FatalExitError
+import org.apache.kafka.common.metadata.{AccessControlRecord, BrokerRecord, ConfigRecord, IsrChangeRecord, PartitionRecord, TopicRecord}
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.raft.RaftMessage
+
+object BrokerMetadataEventManager {
+  val BrokerMetadataEventThreadName = "broker-metadata-event-thread"
+  val EventQueueTimeMetricName = "EventQueueTimeMs"
+  val EventQueueSizeMetricName = "EventQueueSize"
+}
+
+sealed trait Event {}
+
+final case object StartupEvent extends Event
+final case class MetadataEvent(metadataMessage: RaftMessage) extends Event
+/*
+ * ShutdownEvent is necessary for the case when the event thread is blocked in queue.take() -- this event wakes it up.
+ * Otherwise, this event has no other semantic meaning, it is not guaranteed to be seen by the processing thread,
+ * and it cannot be relied upon to initiate any other functionality; it exists solely for the wakeup case mentioned.
+ */
+final case object ShutdownEvent extends Event
+
+class QueuedEvent(val event: Event, val enqueueTimeMs: Long) {
+  override def toString: String = {
+    s"QueuedEvent(event=$event, enqueueTimeMs=$enqueueTimeMs)"
+  }
+}
+
+trait MetadataEventProcessor {
+  def processStartup(): Unit
+  def process(metadataMessage: RaftMessage): Unit
+  // note the absence of processShutdown(); see ShutdownEvent above for why it does not exist.
+}
+
+class BrokerMetadataEventManager(replicaManager: ReplicaManager,
+                                 groupCoordinator: GroupCoordinator,
+                                 metadataCache: MetadataCache,
+                                 quotas: QuotaManagers,
+                                 clusterId: String,
+                                 time: Time,
+                                 overrideProcessor: Option[MetadataEventProcessor] = None,
```

Comment:
Rather than having this, how about having a map from record IDs to message handlers?  This would make unit testing the queue itself easier and move the logic for actually handling the events into separate classes which could also be unit tested in a simpler way.

Reply:
Agree, will work on separating out the logic to ease testing

Reply:
Ok, I now accept `processors: util.EnumMap[MetadataRecordType, ApiMessageProcessor]`